types/wlr_input_method_v2.c

```diff
@@ -20,10 +24,24 @@ static struct wlr_input_method_v2 *input_method_from_resource(
 	return wl_resource_get_user_data(resource);
 }
 
+static struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_input_method_keyboard_grab_v2_interface,
+		&im_keyboard_grab_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void im_keyboard_grab_destroy(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab);
+
 static void input_method_destroy(struct wlr_input_method_v2 *input_method) {
 	wlr_signal_emit_safe(&input_method->events.destroy, input_method);
 	wl_list_remove(wl_resource_get_link(input_method->resource));
 	wl_list_remove(&input_method->seat_destroy.link);
+	if (input_method->im_keyboard_grab != NULL) {
+		wl_resource_destroy(input_method->im_keyboard_grab->resource);
```

Comment:
The compositor cannot destroy the client's resources, except in the `destroy` request handler. Instead, the compositor can make the resource inert.

For more info, see `CONTRIBUTING.md`.

Reply:
Done.